### DISCUSSION TITLE
Compiled `mssql` CREATE TABLE statement does not match DDL emitted by `Base.metadata.create_all`

### DISCUSSION THREAD

**Discussion by thesketh** [sqlalchemy/sqlalchemy#9044] [Usage Questions] [answered]
My ultimate goal is to get the CREATE TABLE statements for each table in a declarative base. I want to do this so I can show some less Python-proficient people in our team that the data models we are creating are equivalent to the old handwritten SQL statements we were using.

When attempting to emit a SQL CREATE TABLE statement for a table, the statement does not match the statement executed by `Base.metadata.create_all`. I have a minimal working example using a Date column which is rendered as `DATETIME` using the former method:

```python
import logging

from sqlalchemy.orm import declarative_base
from sqlalchemy import create_engine, Column, Date, Integer
from sqlalchemy.schema import CreateTable

ENGINE = create_engine("mssql+pyodbc://SA:hunter2!@localhost:1433/master?driver=ODBC+Driver+18+for+SQL+Server&TrustServerCertificate=yes")
Base = declarative_base()
# Enable statement output for engine.
logging.basicConfig()
logging.getLogger("sqlalchemy.engine").setLevel(logging.INFO)


class DatabaseTable(Base):
    __tablename__ = "DatabaseTable"

    Index = Column(Integer, primary_key=True)
    DateColumn = Column(Date)


if __name__ == "__main__":
    # Statement as emitted by compiling the CREATE TABLE statement:
    create_table = CreateTable(DatabaseTable.__table__)
    # Compiling this statement with `dialect=...` produces the same results as passing the engine.
    create_table_statement = str(create_table.compile(ENGINE)).strip()
    print(create_table_statement)  # Emitted statement is wrong, `DateColumn` should be `DATE`.
    # CREATE TABLE [DatabaseTable] (
    #     [Index] INTEGER NOT NULL IDENTITY,
    #     [DateColumn] DATETIME NULL,
    #     PRIMARY KEY ([Index])
    # )

    # Statement which is actually emitted in logs by `Base.metadata.create_all`:
    Base.metadata.create_all(ENGINE)
    # CREATE TABLE [DatabaseTable] (
    #         [Index] INTEGER NOT NULL IDENTITY,
    #         [DateColumn] DATE NULL,
    #         PRIMARY KEY ([Index])
    # )

``` 

How can I get the CREATE TABLE which would be emitted for a specific SQL Server table without actually creating the table and scraping the logs? Is this discrepancy a bug in SQLAlchemy? I'm currently using version 1.4.29, the discrepancy still exists in SQLAlchemy 1.4.45 and 1.3.24.

**Comment by zzzeek** [comment, marked as answer]
the SQL Server dialect has conditional behavior for this particular type because prior to SQL Server 2008 there was no DATE type.   Because the engine in the example hasn't actually established a connection, it is assuming SQL Server < 2008.   So in the most minimal case, adding this line:

    ENGINE.connect().close()

will initialize the dialect and get you DATE instead.    There's a small note regarding this (would not expect one to have seen this easily) at https://docs.sqlalchemy.org/en/14/dialects/mssql.html#date-time-handling .

When you want an exact type the best approach is to use the UPPERCASE type directly:

```py
from sqlalchemy import DATE

class DatabaseTable(Base):
    __tablename__ = "DatabaseTable"

    Index = Column(Integer, primary_key=True)
    DateColumn = Column(DATE)
```

to do this in a dialect-agnostic way, use [with_variant](https://docs.sqlalchemy.org/en/14/core/type_basics.html#using-uppercase-and-backend-specific-types-for-multiple-backends):

```py
from sqlalchemy import DATE

class DatabaseTable(Base):
    __tablename__ = "DatabaseTable"

    Index = Column(Integer, primary_key=True)
    DateColumn = Column(Date().with_variant(DATE, "mssql"))

```

**Reply by thesketh** [reply]
Nice, that should give me enough to go on for now, thanks for your quick response! For future reference, is there a way to do this when passing a dialect rather than an engine? That would be quite useful for unit tests.

**Reply by zzzeek** [reply]
we dont have a constructor flag for this so you'd need to set it manually

```py
if __name__ == "__main__":

    dialect = mssql.dialect()
    dialect.server_version_info = (14, )

    create_table = CreateTable(DatabaseTable.__table__)
    create_table_statement = create_table.compile(dialect=dialect)
    print(create_table_statement)


```

**Reply by thesketh** [reply]
Nice, that's perfect! Thanks again for your help.
